### DISCUSSION TITLE
Get mouse position reference to position Node

### DISCUSSION THREAD

**Discussion by Kettss1** [xyflow/xyflow#3206] [Q&A]
So, I created a custom function to add Nodes to my flow and my idea is to create a Node on double click on the mouse position. I followed every instruction that I found on forums and issues about the position of the mouse with "project" but it does not seems to work for me. 

Here is the code (very simple):

```
  const onMouseMove = (event: MouseEvent) => {
    if(rfDomNode) {
      const bounds = rfDomNode.getBoundingClientRect();
      mousePosRef.current = {
        x: event.clientX - bounds.x,
        y: event.clientY - bounds.y,
      }
    }
  }

// I register the mouse event with 'onMouseMove' from ReactFlow
       onMouseMove={(event) => onMouseMove(event.nativeEvent)}
```
Where rfDomNode is the domNode from the state.
The result that I get is the node being rendered away from the mouse position. Can someone help me?

**Comment by bcakmakoglu** [comment]
You're probably looking for [`project`](https://reactflow.dev/docs/api/react-flow-instance/#project).

Take a look at the [drag and drop example](https://reactflow.dev/docs/examples/interaction/drag-and-drop/) from the docs

```ts
const reactFlowBounds = reactFlowWrapper.current.getBoundingClientRect();

// translate mouse position to graph position
const position = reactFlowInstance.project({
  x: event.clientX - reactFlowBounds.left,
  y: event.clientY - reactFlowBounds.top,
});
```

**Reply by Kettss1** [reply]
I actually tried to use it too, but `left` and `top` are always 0.
![image](https://github.com/wbkd/react-flow/assets/89028159/8ec2a5e7-711b-4289-951b-fea2d2754106)

**Reply by bcakmakoglu** [reply]
Yeah that may very well be, what's the problem though? 😄

Did you actually use the `project` function to get the right coordinates? What happened? Is there a wrong position on drop? Can you give an example? A reproduction? :)

**Reply by Kettss1** [reply]
yes, I used `project` when specifying the position for a new node:
`const { x, y } = project({ x: mousePosRef.current.x, y: mousePosRef.current.y });`

I'm following the [Copy and Paste](https://pro.reactflow.dev/examples/copy-paste) example, but when I use the shortcut (ctrl + v), the Node ends up far away from the mouse.

**Reply by Kettss1** [reply]
I'll try and create a reproduction of my repo.

**Reply by bcakmakoglu** [reply]
If it's about a pro example, you're better off contacting the dev team at `info@reactflow.dev` directly

**Reply by Kettss1** [reply]
thank you for all your help

**Comment by moklick** [comment]
Could you show the whole function that you wrote to add the node?

**Reply by Kettss1** [reply]
Hey! Thank you for your response. I made a shorter version of my app on this repo: 
https://github.com/Kettss1/reactflow-tests
I tested it and the error is happening. Thank you again.

**Reply by moklick** [reply]
Things are not working here, because you are not passing the needed dependencies to the useCallback dependency arrays. 

- https://github.com/Kettss1/reactflow-tests/blob/main/src/components/board/board.tsx#L118 `project` is missing
- https://github.com/Kettss1/reactflow-tests/blob/main/src/components/board/board.tsx#L218 `dbclickFunction` and `keyboardShortcutsFunction` are missing

You always need to make sure that all things that could change over time are getting passed to the dep array.

**Reply by Kettss1** [reply]
Sorry for the late response. I actually tried that, but this created a "maximum update depth exceeded" error. I'm testing some stuff out but would appreciate your help. Thank you in advance.

**Reply by ghost** [reply]
@Kettss1 Hi, did you found a solution?

**Comment by dbfr3qs** [comment]
I think `project` might be deprecated - I couldn't find it anywhere in the documentation. I solved this using [ReactFlowInstance.screenToFlowPosition](https://reactflow.dev/api-reference/types/react-flow-instance#screen-to-flow-position) which is easily accessible via the [useReactFlow](https://reactflow.dev/api-reference/hooks/use-react-flow) hook (which is how they're doing it in the drag and drop [demo](https://reactflow.dev/examples/interaction/drag-and-drop)).
